src/librustc_privacy/lib.rs

```diff
@@ -1670,6 +1706,10 @@ fn privacy_access_levels<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     let krate = tcx.hir().krate();
     let empty_tables = ty::TypeckTables::empty(None);
 
+    let public_crates: FxHashSet<CrateNum> = tcx.sess.opts.extern_public.iter().flat_map(|c| {
+        tcx.crates().iter().find(|&&krate| &tcx.crate_name(krate) == c).cloned()
```

Comment:
This is a bit suspicious.
This will collect crates that has names from `--extern-public` *in their metadata*.
So names in `--extern-public` don't have direct correspondence to names passed with `--extern name` or names in `extern crate name`.
I'm not sure what the intent of the RFC is here.

I would expect `--extern-public="my_crate"` to mark crate passed with `--extern my_crate=...` as public.
Not sure.

Reply:
Ideally, the public marker should be somehow attached to the `--extern`  options themselves, e.g.
```sh
rustc --extern crate1=PATH1 --extern-public crate2=PATH2
# or
rustc --extern crate1=PATH1 --extern crate2:public=PATH2
# or something similar
```
This way config from `Cargo.toml` would translate directly and precisely into command line options.
(This can also work with `PATH`-less `--extern`.)

Reply:
@petrochenkov: How would we distinguish between "no public deps, but run the leak checking" and "don't run leak checking"? With this approach, we'd need an extra flag to indicate the first case.

Reply:
@Aaron1011 
To clarify, I don't suggest to implement the option changes right now, what's implemented right now is good enough for experimenting with the feature on nightly.
I mostly expected some feedback from e.g. @alexcrichton.
The precise semantics need to be figured out before stabilization though.

If `public` annotations are moved into existing `--extern` options, then yes, perhaps a new option would be needed to enable/disable leak checking.

Reply:
Alternatively, the default could be flipped for `rustc` compared to Cargo, `--extern my_crate=PATH` would still mean public, and Cargo would generate `--extern my_crate:priv=PATH` for private dependencies instead.

Reply:
I personally like the idea of `--extern foo=bar` in addition to `--extern-public foo=bar`. I definitely think that the name `foo` in `--extern foo=path/to/libfoo.rlib` should be connected to the CLI option for "is `foo` private or not?" as that's how Cargo will be passing flags. 

I think the rollout plan for Cargo is still somewhat up in the air, but having `--extern` default to "public" seems like a good place to start. In that case we may want to just add `--extern-private` as a flag perhaps?